Content.Server/GameObjects/EntitySystems/Weapon/Melee/StunbatonSystem.cs

```diff
@@ -0,0 +1,185 @@
+﻿using Content.Server.GameObjects.Components.Items.Storage;
+using Content.Server.GameObjects.Components.Mobs;
+using Content.Server.GameObjects.Components.Power;
+using Content.Server.GameObjects.Components.Weapon.Melee;
+using Content.Shared.Audio;
+using Content.Shared.GameObjects.Components.Power;
+using Content.Shared.GameObjects.EntitySystems;
+using Content.Shared.GameObjects.EntitySystems.ActionBlocker;
+using Content.Shared.Interfaces;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.Audio;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Log;
+using Robust.Shared.Player;
+using Robust.Shared.Random;
+
+namespace Content.Server.GameObjects.EntitySystems.Weapon.Melee
+{
+    public class StunbatonSystem : EntitySystem
+    {
+        [Dependency] private readonly IRobustRandom _robustRandom = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<StunbatonComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<StunbatonComponent, MeleeHitEvent>(OnMeleeHit);
+            SubscribeLocalEvent<StunbatonComponent, UseInHandEvent>(OnUseInHand);
+            SubscribeLocalEvent<StunbatonComponent, ThrowCollideEvent>(OnThrowCollide);
+            SubscribeLocalEvent<StunbatonComponent, PowerCellChangedEvent>(OnPowerCellChanged);
+            SubscribeLocalEvent<StunbatonComponent, InteractUsingEvent>(OnInteractUsing);
+            SubscribeLocalEvent<StunbatonComponent, ExaminedEvent>(OnExamined);
+        }
+
+        private void OnAfterInteract(EntityUid uid, StunbatonComponent comp, AfterInteractEvent args)
+        {
+            if (!comp.Activated || args.Target == null || comp.Cell == null)
+                return;
+
+            if (!comp.Cell.TryUseCharge(comp.EnergyPerUse))
+                return;
+
+            StunEntity(args.Target, comp);
+        }
+
+        private void OnMeleeHit(EntityUid uid, StunbatonComponent comp, MeleeHitEvent args)
+        {
+            if (!comp.Activated || args.HitEntities.Count == 0 || comp.Cell == null)
```

Comment:
You should get the cell component thing here with `ComponentManager`, and check that manually